Automattic/wp-calypso
# Framework: Upgrade to React 0.14

This pull request seeks to upgrade Calypso from React 0.13.3 to React 0.14.3. This requires a significant number of changes, particularly related to the split between `react` and `react-dom` packages ([see full change log](https://facebook.github.io/react/blog/2015/10/07/react-v0.14.html)).

The upgrade results in many warnings during testing and development. These include advisements to use `react-addons-` packages in place of `react/addons` and against rendering to the `document.body` element (common in tests). The purpose of this pull request is not to resolve these warnings and assumes that they'll be addressed separately.

However, warnings on use of `getDOMNode` __were__ included in this pull request, as they are pervasive enough to be considered obnoxious in using the application in development environments. All instances of `getDOMNode` have been replaced with direct `ref` usage for DOM elements (since a `ref` is the DOM node reference in React 0.14) or `React.findDOMNode` (when the element is a React component or references the `this` instance directly).

__Testing instructions:__

Run `make test` from the root directory and verify that all tests pass (with warnings).

Ensure that regressions should be found in navigating the application in the browser.

__Code verification:__

Ensure that no [`ReactDOM` methods](https://facebook.github.io/react/docs/top-level-api.html#reactdom) are used with the base `React` package.

Verify that all removals of `getDOMNode` apply only to DOM elements.
